apps/hexagon_benchmarks/process.h

```diff
@@ -0,0 +1,482 @@
+#endif
+
+
+enum bmark_run_mode_t {
+    hvx64 = 1,
+    hvx128 = 2,
+    cpu = 3
+};
+
+template <typename T>
+T clamp(T val, T min, T max) {
+    if (val < min)
+        return min;
+    if (val > max)
+        return max;
+    return val;
+}
+
+typedef int (*pipeline2)(buffer_t *, buffer_t *);
+typedef int (*pipeline3)(buffer_t *, buffer_t *, buffer_t *);
+
+struct PipelineDescriptorBase {
+    virtual void init() = 0;
+    virtual const char * name() = 0;
+    virtual int run(bmark_run_mode_t mode) = 0;
+    virtual bool verify(int W, int H) = 0;
+    virtual bool defined() = 0;
+};
+
+class Conv3x3a16Descriptor : public PipelineDescriptorBase {
+    Halide::Runtime::Buffer<uint8_t> u8_in, u8_out;
+    Halide::Runtime::Buffer<int8_t> i8_mask;
+
+public:
+    Conv3x3a16Descriptor(int W, int H) : u8_in(nullptr, W, H, 2),
+                                         u8_out(nullptr, W, H, 2),
+                                         i8_mask(nullptr, 3, 3, 2) {}
+
+    void init() {
+        u8_in.device_malloc(halide_hexagon_device_interface());
+        u8_out.device_malloc(halide_hexagon_device_interface());
+        i8_mask.device_malloc(halide_hexagon_device_interface());
+
+        u8_in.for_each_value([&](uint8_t &x) {
+            x = static_cast<uint8_t>(rand());
+        });
+        u8_out.fill(0);
+
+        i8_mask(0, 0) = 1;
+        i8_mask(1, 0) = -4;
+        i8_mask(2, 0) = 7;
+
+        i8_mask(0, 0) = 2;
+        i8_mask(1, 0) = -5;
+        i8_mask(2, 0) = 8;
+
+        i8_mask(0, 0) = 3;
+        i8_mask(1, 0) = -6;
+        i8_mask(2, 0) = 7;
+    }
+
+
+    const char *name() { return "conv3x3a16"; }
+
+    bool defined() {
+#ifdef CONV3X3A16
+        return true;
+#else
+        return false;
+#endif
+    }
+
+    bool verify(const int W, const int H) {
+        u8_out.for_each_element([&](int x, int y) {
+            int16_t sum = 0;
+            for (int ry = -1; ry <= 1; ry++) {
+                for (int rx = -1; rx <= 1; rx++) {
+                    sum += static_cast<int16_t>(u8_in(clamp(x+rx, 0, W-1), clamp(y+ry, 0, H-1)))
+                                                * static_cast<int16_t>(i8_mask(rx+1, ry+1));
+                }
+            }
+            sum = sum >> 4;
+            sum = clamp<int16_t>(sum, 0, 255);
+            uint8_t out_xy = u8_out(x, y);
+            if (sum != out_xy) {
+                printf("Conv3x3a16: Mismatch at %d %d : %d != %d\n", x, y, out_xy, sum);
+                abort();
+            }
+        });
+        return true;
+    }
+
+    int run(bmark_run_mode_t mode) {
+#ifdef CONV3X3A16
+        if (mode == bmark_run_mode_t::hvx64) {
+            return conv3x3a16_hvx64(u8_in, i8_mask, u8_out);
+        } else if (mode == bmark_run_mode_t::hvx128) {
+            return conv3x3a16_hvx128(u8_in, i8_mask, u8_out);
+        } else if (mode == bmark_run_mode_t::cpu) {
+            return conv3x3a16_cpu(u8_in, i8_mask, u8_out);
+        }
+#endif
+        return 0;
```

Comment:
This should probably return non-zero to indicate error, as a pipeline invocation would have done. I'd say return 1, because halide errors are negative.

Same for the other descriptors.